In filter_topo added extrapolation to halo points

RatkoVasic-NOAA opened the pull request:
This commit is related to the issue #105
Fields' geolat_c and geolon_c halo points are used to calculate dxc and dyc, but halo points are not initialized for regional standalone case. I added extrapolation for regional standalone similar way to regional nest and global application.

RatkoVasic-NOAA committed:
Add extrapolation for geolat and geolon to halo points. This commit references issue #105

GeorgeGayno-NOAA commented:
Thanks @RatkoVasic-NOAA. Does routine 'read_topo_file` have the same problem. What is in the halo region of arrays 'oro' and 'mask'?

RatkoVasic-NOAA commented:
> Thanks @RatkoVasic-NOAA. Does routine 'read_topo_file` have the same problem. What is in the halo region of arrays 'oro' and 'mask'?

The only arrays (_in regional standalone mode_) that were using halo points are geolat_c and geolon_c. All other arrays are not using halo points.

GeorgeGayno-NOAA commented on sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90:
```diff
@@ -1144,6 +1147,42 @@ subroutine fill_cubic_grid_halo(data, data2, halo, ioff, joff, sign1, sign2)
+  !> This routine extrapolate geolat_c and geolon_c halo points for the
+  !! regional standalone grid. Halo points are needed for dxc and dyc
+  !! calculation.
+  !!
```
Replace the ??? with the variable definition. The ??? was a placeholder so we could get a basic doxygen capability in this code.  And use your full name and work center - "NCEP/EMC" - after 'author'.

GeorgeGayno-NOAA commented on sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90:
```diff
@@ -1144,6 +1147,42 @@ subroutine fill_cubic_grid_halo(data, data2, halo, ioff, joff, sign1, sign2)
+  !! @param[in] data  ???
+  !! @param[in] halo ???
+  !! @author Ratko
+  subroutine fill_regional_halo(data, halo)
```
We are required to do unit tests for all new routines. These tests run under github actions. I can help you write the test. Just ask.

RatkoVasic-NOAA replied:
I'm adding descriptions. I see everywhere @parm[in], does doxygen recognize [inout]?

RatkoVasic-NOAA replied:
Yes, I will need help to create test.

GeorgeGayno-NOAA replied:
> I'm adding descriptions. I see everywhere @parm[in], does doxygen recognize [inout]?

Yes.

RatkoVasic-NOAA replied:
> > I'm adding descriptions. I see everywhere @parm[in], does doxygen recognize [inout]?
> 
> Yes.

Oh, I see in manual, syntax is [in,out].

GeorgeGayno-NOAA commented:
@RatkoVasic-NOAA I tested your branch at b602411 on my canned test case. I compiled your branch on Hera with CMAKE_BUILD_TYPE equal to "Release" and "Debug". I got big differences along the halo.  I will provide details.

GeorgeGayno-NOAA commented:
Here are the differences between "Debug" and "Release" mode for my test case:

```
+ nccmp -dmqfS oro.C424.tile7.nc oro.C424.tile7.nc.ratko.intel.O3
Variable  Group Count     Sum AbsSum     Min     Max   Range     Mean  StdDev
orog_filt /        55 1.90058 1901.6 -239.36 239.794 479.154 0.034556 70.2504
```

I visualized the files. The differences are confined to the halo.


If you do not already know, you can change the compile type using this variable: `-DCMAKE_BUILD_TYPE=Debug`

RatkoVasic-NOAA commented:
I'll test it.

RatkoVasic-NOAA committed:
Add comments for doxygen.

RatkoVasic-NOAA committed:
Fix bugs in FV3_zs_filter and two_delta_filter.

GeorgeGayno-NOAA commented on sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90:
```diff
@@ -1295,6 +1334,8 @@ subroutine two_delta_filter(is, ie, js, je, isd, ied, jsd, jed, npx, npy, ntiles
     integer:: i,j, nt, t
     integer:: is1, ie2, js1, je2
 
+    a2 = 0.
```
@RatkoVasic-NOAA I might add a comment here about this being a simple fix and your suggestion for a better fix.

RatkoVasic-NOAA replied:
Let's hold on for the moment, I'm looking into better solution.

GeorgeGayno-NOAA replied:
Great. By the way, the random behavior seems to have disappeared. I tested 06cc530 using Intel/GNU, Release/Debug compilations. My canned case gave identical results for all four tests. So I think you found all the problem arrays.

RatkoVasic-NOAA committed:
Fix oro and mask fields on halo regions in regional application.

GeorgeGayno-NOAA committed:
Move new routine 'ftst_fill_regional_halo' to utils.F90
to allow for using it in unit tests. Create simple unit
test for that routine.

Fixes #105

GeorgeGayno-NOAA committed:
Comment out call to routine in new unit test.

Fixes #105

RatkoVasic-NOAA committed:
Fixed coefficient a1 in subroutine two_delta_filter. Corners were used, but not initialized.

RatkoVasic-NOAA committed:
Add test for routine fill_regional_halo.

RatkoVasic-NOAA committed:
Fix 'module use' in test file.

GeorgeGayno-NOAA commented:
@RatkoVasic-NOAA I see that your unit test was failing with a memory leak error. In your unit test, try deallocating "testdata" at the end of program execution.

RatkoVasic-NOAA committed:
Add deallocate.

GeorgeGayno-NOAA commented:
It works!

GeorgeGayno-NOAA commented:
@JeffBeck-NOAA may want to run a quick test with the regional workflow.

@RatkoVasic-NOAA is this ready to merge?

RatkoVasic-NOAA commented:
> 
> @RatkoVasic-NOAA is this ready to merge?

I replied by email, I guess it didn't go through.
Answer is yes, it is ready. Is there anything I should do more for this PR?

GeorgeGayno-NOAA commented:
> > @RatkoVasic-NOAA is this ready to merge?
> 
> I replied by email, I guess it didn't go through.
> Answer is yes, it is ready. Is there anything I should do more for this PR?

The 'grid_gen' regression tests will need to be run. Since results will likely change for the regional tests, new baseline files will need to be generated. @RatkoVasic-NOAA can you run the tests on Hera or WCOSS?

GeorgeGayno-NOAA commented:
And merge to your branch the latest changes from 'develop'.

RatkoVasic-NOAA committed:
Merge branch 'develop' into filter_topo

GeorgeGayno-NOAA commented:
> > > @RatkoVasic-NOAA is this ready to merge?
> > 
> > 
> > I replied by email, I guess it didn't go through.
> > Answer is yes, it is ready. Is there anything I should do more for this PR?
> 
> The 'grid_gen' regression tests will need to be run. Since results will likely change for the regional tests, new baseline files will need to be generated. @RatkoVasic-NOAA can you run the tests on Hera or WCOSS?

Thanks. I have updated the baseline data on all machines.
